### DISCUSSION TITLE
Add support for RelationShipView

### DISCUSSION THREAD

**Discussion by n2ygk** [jokiefer/drf-spectecular-json-api#20] [Ideas]
See discussion at https://github.com/django-json-api/django-rest-framework-json-api/discussions/1082.  Moving that discussion here. I am trying to convert from using the built-in deprecated DRF & DJA OAS 3.x support to using drf-spectacular and drf-spectacular-jsonapi.

I'm testing against an existing project's `spectacular` branch: https://github.com/columbia-it/django-jsonapi-training/tree/spectacular

I had RelationShipViews which caused asserts to fail like this:
```
AssertionError: Expected view CourseTermRelationshipView to be called with a URL keyword argument named "pk". Fix your URL conf, or set the `.lookup_field` attribute on the view correctly.
```

In removing all the RelationShipViews from my urls.conf, I was able to generate a good OAS schema, attached as [drf-spectacular-jsonapi-no-relationships-yaml.txt](https://github.com/user-attachments/files/16367509/drf-spectacular-jsonapi-no-relationships-yaml.txt). It is generated with the following management command in the [spectacular](https://github.com/columbia-it/django-jsonapi-training/tree/spectacular) branch of my project:
```
(venv) django-jsonapi-training$ python manage.py spectacular --file drf-spectacular-jsonapi-no-relationships-yaml.txt
/Users/ac45/src/django-jsonapi-training/venv/lib/python3.12/site-packages/drf_spectacular_jsonapi/schemas/utils.py:14: UserWarning: Can't resolve primary key for non model serializers.
  warn(message="Can't resolve primary key for non model serializers.")
/Users/ac45/src/django-jsonapi-training/myapp/views.py:210: Warning [CourseTermViewSet]: could not derive type of path parameter "related_field" because model "myapp.models.CourseTerm" contained no such field. Consider annotating parameter with @extend_schema. Defaulting to "string".
/Users/ac45/src/django-jsonapi-training/myapp/views.py:186: Warning [CourseViewSet]: could not derive type of path parameter "related_field" because model "myapp.models.Course" contained no such field. Consider annotating parameter with @extend_schema. Defaulting to "string".
/Users/ac45/src/django-jsonapi-training/myapp/views.py:298: Warning [InstructorViewSet]: could not derive type of path parameter "related_field" because model "myapp.models.Instructor" contained no such field. Consider annotating parameter with @extend_schema. Defaulting to "string".
/Users/ac45/src/django-jsonapi-training/myapp/views.py:226: Warning [PersonViewSet]: could not derive type of path parameter "related_field" because model "myapp.models.Person" contained no such field. Consider annotating parameter with @extend_schema. Defaulting to "string".
Warning: operationId "course_terms_retrieve" has collisions [('/v1/course_terms/{id}/', 'get'), ('/v1/course_terms/{id}/{related_field}/', 'get')]. resolving with numeral suffixes.
Warning: operationId "courses_retrieve" has collisions [('/v1/courses/{id}/', 'get'), ('/v1/courses/{id}/{related_field}/', 'get')]. resolving with numeral suffixes.
Warning: operationId "instructors_retrieve" has collisions [('/v1/instructors/{id}/', 'get'), ('/v1/instructors/{id}/{related_field}/', 'get')]. resolving with numeral suffixes.
Warning: operationId "people_retrieve" has collisions [('/v1/people/{id}/', 'get'), ('/v1/people/{id}/{related_field}/', 'get')]. resolving with numeral suffixes.

Schema generation summary:
Warnings: 8 (8 unique)
Errors:   0 (0 unique)
```
You'll note the warnings about non-model serializers as well as the `related_field` warnings.

Compare this to [openapi-yaml.txt](https://github.com/user-attachments/files/16367510/openapi-yaml.txt) which was generated using the deprecated DJA OAS support plus some extensions I'd implemented. It is generated using this management command in the [master](https://github.com/columbia-it/django-jsonapi-training/tree/master) branch:
```
python manage.py generateschema  --generator_class myapp.schemas.SchemaGenerator --file openapi-yaml.txt
```

I believe the issue is that `class RelationshipView(generics.GenericAPIView)` is not based on a `ModelViewSet` as is typically used. So I think some enhancements are needed to support GenericAPIView.

**Comment by n2ygk** [comment]
The OAS schema generated by `drf-spectecular-json-api` looks a lot more complete and well-designed than what was coming out of DJA!

**Reply by n2ygk** [reply]
@jokiefer I did some comparisons of the API vs. the OAS schema generated. See [my latest draft docs](https://columbia-it-django-jsonapi-training--29.org.readthedocs.build/en/29/documenting-api/#new-using-drf-spectacular-jsonapi-as-the-schema-generator)

TL;DR:
* RelationshipViews break `manage.py spectacular`
* HyperlinkedModelSerializer not supported: `links` not implemented.
* Meta objects missing (minor but still something preventing full fidelity).
* Parameters support is really nice!

I've only done some cursory investigation so far.

**Reply by jokiefer** [reply]
wow, thanks for your time investigation on that. I will dig deeper in a few days to fix all that issues.

**Reply by n2ygk** [reply]
@jokiefer let me know if/how I can help

**Reply by jokiefer** [reply]
@n2ygk In my POV the problem is, that `RelationShipView`s are a special case of generic views in the json:api context. This means that there could by different kind of resource types which are related to a `parent` resource type. So i think we need to analyze the view in that way, so that we traverse all possible related resource types, and add this schema definitions to the openapi schema. So the `RelationShipView` is converted to the openapi context something like that:

```
{
  type: "object",
  properties: {
    data: {
      anyOf: 
        - Song
    }
  }
}

Song: {
  type: "object",
      required: ["type", "id"],
      properties: {
        type: {allOf: [Song]},
        id: {type: "string", format: "uuid"}
      }
}
```

And that for all possible related resource types.

See also the current changes: https://github.com/jokiefer/drf-spectecular-json-api/compare/main...20_add_relation_ship_view

**Reply by n2ygk** [reply]
@jokiefer part of the resource name for the relationship view indicates the `<related_field>`. Not sure if that helps.

https://github.com/columbia-it/django-jsonapi-training/blob/master/training/urls.py#L44-L48
```
# TODO: Is there a Router than can generate these for me? If not, create one.
    # course relationships:
    path('v1/courses/<pk>/relationships/<related_field>/',
        views.CourseRelationshipView.as_view(),
        name='course-relationships'),
```

See https://github.com/columbia-it/django-jsonapi-training/blob/master/docs/schemas/openapi.yaml#L2572 for what's currently being generated using legacy DRF/DJA schema support.

**Comment by jokiefer** [comment]
[v0.5.0](https://pypi.org/project/drf-spectacular-jsonapi/0.5.0/) adds support for RelationshipViews
